### ISSUE TITLE
00_eda

### ISSUE THREAD

**Issue by mezcolantriz** [4GeeksAcademy/afernandez-https-github.com-4GeeksAcademy-algorithm-optimization-project-machine-learning#1]
@consultingafernandez-hub 

### **1. LEVE.. Imports esparcidos**

Tienes imports en varias celdas durante todo el notebook. **Ponlos todos al principio en la celda de imports juntos y quita el resto**

---

#### PRIMERO LEE EL PROBLEMA, luego ten en cuenta la siguiente issue antes de modificar ya que están relacionadas.
**El problema de data leakage es que los datos del test escalados antes el split digamos que "ven" parte del test, lo que se llamaría una fuga de datos. Tienes:**

```python
# PASO 1: ESCALAS TODO EL DATASET ANTES DEL SPLIT no es correcto! 
scaler = MinMaxScaler()
scaled_features = scaler.fit_transform(total_data[num_variables])  # ← FIT en TODO

# PASO 2: DESPUÉS hace el split 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, ...)
```

**¿Por qué es problema?**

El scaler se ajustó usando:
- **Mínimo global** = mínimo de TODO el dataset (incluyendo test)
- **Máximo global** = máximo de TODO el dataset (incluyendo test)

**Resultado:** El test set influye en cómo se normalizan los datos. El modelo "ve" información del futuro.

**La forma correcta:**

```python
# PASO 1: SPLIT PRIMERO
X = total_data.drop("price", axis=1)
y = total_data["price"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# PASO 2: En issue de más abajo NO MODIFIQUES AÚN
# PASO 3: FIT SCALER SOLO EN TRAIN
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train)  # ← FIT solo en train

# PASO 4: TRANSFORM TEST CON MISMO SCALER (SIN REFIT)
X_test_scaled = scaler.transform(X_test)  # ← No fit, solo transform

# PASO 5: Convertir a DataFrame
X_train_scaled = pd.DataFrame(X_train_scaled, columns=X_train.columns)
X_test_scaled = pd.DataFrame(X_test_scaled, columns=X_test.columns)

X_train_scaled["price"] = y_train.values
X_test_scaled["price"] = y_test.values

```

---
### **3. IMPORTANTE. UTILIZAS FACTORIZE() como encoding final**

Veamos qué ocurre actualmente:

1. **Matriz de correlación**: Usas `factorize()` para visualizar, lo que **modifica permanentemente `total_data`**:
   ```python
   total_data["room_type"] = pd.factorize(total_data["room_type"])[0]
   total_data["neighbourhood_group"] = pd.factorize(total_data["neighbourhood_group"])[0]
   ```
> Factorize está bien para visualizar relaciones pero es muy débil para dejarlo como transformación final a numéricas para el modelo. Mejor usar un encoding apropiado.

2. **Escalado**: Luego aplicas scaler ANTES del split # VER LA ISSUE DE MÁS ARRIBA:
   ```python
   scaler.fit_transform(total_data[num_variables])  # Escalas sobre TODO el dataset
   ```

3. **Después**: Haces `train_test_split()` sobre datos ya escalados (data leakage)

---

**El flujo CORRECTO debería ser:**

```python
# 1 - EDA: Usar factorize SOLO para visualización (sin modificar total_data)
# Se puede crear una copia temporal para el heatmap o quitarlos después ejemplo con primera opción:
temp_data = total_data.copy()
temp_data["room_type"] = pd.factorize(temp_data["room_type"])[0]
temp_data["neighbourhood_group"] = pd.factorize(temp_data["neighbourhood_group"])[0]
# ... graficar heatmap con temp_data ...

# 2- SPLIT PRIMERO (en datos sin escalar - problema del punto 2)
X = total_data.drop("price", axis=1)
y = total_data["price"]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y)

# 3 - ENCODING REAL después del split
# One-hot encoding o label encoding
X_train_encoded = pd.get_dummies(X_train, columns=["room_type", "neighbourhood_group"])
X_test_encoded = pd.get_dummies(X_test, columns=["room_type", "neighbourhood_group"])

# 4 - SCALER en training, transform en test
scaler = MinMaxScaler()
X_train_scaled = scaler.fit_transform(X_train_encoded)
X_test_scaled = scaler.transform(X_test_encoded)

# 5- Feature selection (opcional)
selector = SelectKBest(chi2, k=4)
selector.fit(X_train_scaled, y_train)
# ...
```
